smolyay/grid.py

```diff
@@ -2,348 +2,429 @@
-        """
-        grid_points_indexes = None
-        for sum_of_levels in range(dimension,
-                                   dimension+len(self._basis_set.levels)):
-            for composition in generate_compositions(
-                    sum_of_levels, dimension, include_zero=False):
-                index_composition = numpy.array(composition) - 1
-                # generate all combinations of
-                # the arrays along each dimension
-                level_composition_index = [self._basis_set.levels[index]
-                                           for index in index_composition]
-                # skip compositions containing empty levels
-                if any(len(x) == 0 for x in level_composition_index):
-                    continue
-                grid_points_indexes_ = (numpy.array(
-                    numpy.meshgrid(*level_composition_index))
-                    .T.reshape(-1, dimension))
-                if grid_points_indexes is None:
-                    grid_points_indexes = grid_points_indexes_
-                else:
-                    grid_points_indexes = numpy.concatenate(
-                        (grid_points_indexes,
-                         grid_points_indexes_), axis=0)
-        # make integer grids, levels, grid points and basis functions
-        grid_points_indexes = grid_points_indexes.tolist()
-        level_lookup = {point: level for level, points in
-                        enumerate(self._basis_set.levels) for point in points}
-        levels = []
-        grid_points = []
-        grid_points_basis = []
-        for indexes in grid_points_indexes:
-            levels.append([level_lookup[index]
-                       for index in indexes])
-            grid_points.append([self._basis_set.points[index]
-                                for index in indexes])
-            grid_points_basis.append([self._basis_set.basis_functions[index]
-                                      for index in indexes])
-
-        # flatten 1d arrays
-        if dimension == 1:
-            grid_points_indexes = [i[0] for i in grid_points_indexes]
-            grid_points = [pt[0] for pt in grid_points]
-            grid_points_basis = [b[0] for b in grid_points_basis]
-            levels = [l[0] for l in levels]
-
-        return NestedIndexGrid(grid_points_indexes, grid_points,
-                               grid_points_basis, levels)
-
-
-class TensorGridGenerator(IndexGridGenerator):
-    """Create grid points and functions for constructing a surrogate.
-
-    Depending on the dimensionality, points and basis functions provided
-    by :class:`BasisFunctionSet`, the :meth:`__call__` makes full tensor
-    grids.
-    :meth:`__call__`  generates the indexes of
-    grid points (integer grids) depending on dimensionality (full tensor grid).
-    Grid points and their functions are made by replacing the indexes with
-    corresponding point.
-    """
+    @property
+    def number_points(self):
```

Comment:
This would be a length method, following what you did for the unidimensional point sets and belong in the base class.
```suggestion
    def __len__(self):
```